Source/Scene/ModelExperimental/PointCloudAttenuationPipelineStage.js

```diff
@@ -0,0 +1,120 @@
+import Cartesian3 from "../../Core/Cartesian3.js";
+import defaultValue from "../../Core/defaultValue.js";
+import defined from "../../Core/defined.js";
+import OrthographicFrustum from "../../Core/OrthographicFrustum.js";
+
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+import PointCloudAttenuationStageVS from "../../Shaders/ModelExperimental/PointCloudAttenuationStageVS.js";
+import SceneMode from "../SceneMode.js";
+import ModelExperimentalType from "./ModelExperimentalType.js";
+
+/**
+ * Stage to handle point cloud attenuation.
+ *
+ * @namespace PointCloudAttenuationPipelineStage
+ *
+ * @private
+ */
+var PointCloudAttenuationPipelineStage = {};
+PointCloudAttenuationPipelineStage.name = "PointCloudAttenuationPipelineStage"; // Helps with debugging
+
+var scratchAttenuationUniform = new Cartesian3();
+
+/**
+ * This stage does the following:
+ * <ul>
+ *  <li>Add vertex shader code to compute attenuation and update gl_PointSize</li>
+ *  <li>Updates the uniform map to pass in the point cloud attenuation parameters</li>
+ * </ul>
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive
+ * @param {ModelComponents.primitive} primitive The primitive
+ * @param {FrameState} frameState The frame state
+ *
+ * @private
+ */
+PointCloudAttenuationPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  var shaderBuilder = renderResources.shaderBuilder;
+  shaderBuilder.addVertexLines([PointCloudAttenuationStageVS]);
+  shaderBuilder.addDefine(
+    "USE_POINT_CLOUD_ATTENUATION",
+    undefined,
+    ShaderDestination.VERTEX
+  );
+
+  var model = renderResources.model;
+  var pointCloudShading;
+  var content;
+  if (ModelExperimentalType.is3DTiles(model.type)) {
+    content = model.content;
+    pointCloudShading = content.tileset.pointCloudShading;
+  } else {
+    pointCloudShading = model.pointCloudShading;
+  }
+
+  shaderBuilder.addUniform(
+    "vec3",
+    "model_pointCloudAttenuation",
+    ShaderDestination.VERTEX
+  );
+  renderResources.uniformMap.model_pointCloudAttenuation = function () {
+    var scratch = scratchAttenuationUniform;
+
+    // attenuation.x = pointSize
+    scratch.x = pointCloudShading.attenuation
+      ? defaultValue(pointCloudShading.maximumAttenuation, 1.0)
+      : 1.0;
+    scratch.x *= frameState.pixelRatio;
+
+    if (pointCloudShading.attenuation) {
```

Comment:
Remove this, I should check for attenuation when the pipeline stages are selected.